src/DistributedLock.Postgres/PostgresDistributedLock.Extensions.cs

```diff
@@ -0,0 +1,56 @@
+﻿using Medallion.Threading.Internal;
+using System.Data;
+
+namespace Medallion.Threading.Postgres;
+
+public partial class PostgresDistributedLock
+{
+    public static bool TryAcquireWithTransaction(PostgresAdvisoryLockKey key, IDbTransaction transaction, TimeSpan timeout = default, CancellationToken cancellationToken = default)
+    {
+        if (key == null) { throw new ArgumentNullException(nameof(key)); }
+        if (transaction == null) { throw new ArgumentNullException(nameof(transaction)); }
+
+        var connection = new PostgresDatabaseConnection(transaction);
+
+        var handle = DistributedLockHelpers.TryAcquire(PostgresAdvisoryLock.ExclusiveLock, connection, key.ToString(), timeout, cancellationToken);
+
+        return handle != null;
+    }
+
+    public static void AcquireWithTransaction(PostgresAdvisoryLockKey key, IDbTransaction transaction, TimeSpan? timeout = null, CancellationToken cancellationToken = default)
+    {
+        if (key == null) { throw new ArgumentNullException(nameof(key)); }
+        if (transaction == null) { throw new ArgumentNullException(nameof(transaction)); }
+
+        var connection = new PostgresDatabaseConnection(transaction);
+
+        DistributedLockHelpers.Acquire(PostgresAdvisoryLock.ExclusiveLock, connection, key.ToString(), timeout, cancellationToken);
+    }
+
+    public static ValueTask<bool> TryAcquireWithTransactionAsync(PostgresAdvisoryLockKey key, IDbTransaction transaction, TimeSpan timeout = default, CancellationToken cancellationToken = default)
+    {
+        if (key == null) { throw new ArgumentNullException(nameof(key)); }
+        if (transaction == null) { throw new ArgumentNullException(nameof(transaction)); }
+
+        var connection = new PostgresDatabaseConnection(transaction);
+
+        return TryAcquireAsync();
+
+        async ValueTask<bool> TryAcquireAsync()
+        {
+            var handle = await DistributedLockHelpers.TryAcquireAsync(PostgresAdvisoryLock.ExclusiveLock, connection, key.ToString(), timeout, cancellationToken);
```

Comment:
I think we want to call `await handle.DisposeAsync().ConfigureAwait(false);` here with a comment saying that for an externally-owned transaction the release is a noop but we want to dispose proactively to prevent the handle's managed finalizer (see ManagedFinalizerQueue) from running.

Reply:
I am afraid that's not possible because of how the code is written. I actually tried to avoid this exact scenario, since disposal of the handle would have reached the `ReleaseAsync` method in the `PostgresAdvisoryLock` class - where you didn't like the changes I did there. Instead, I am calling the `PostgresAdvisoryLock` class directly in order to acquire the lock and get an object cookie in return, which I ignore. Releasing the handle would have required to go through the `DedicatedConnectionOrTransactionDbDistributedLock` class.